File: project/backend/src/services/analysisEngine.ts
/**
 * VentureWise Feasibility Analysis Engine v2
 * Rule-based, realistic, specific scoring — not random.
 * Uses products/services, competitors, and city data for deep analysis.
 */

import { getCompetitorsForBusiness, type Competitor } from "../data/competitors.js";
import { getCityData, getZonesByBusinessType, type CommercialZone } from "../data/cityData.js";

interface AnalysisInput {
  businessType: string;
  description: string;
  budget: number;
  targetAudience: string;
  productsOrServices: string;
  categoryName: string;
  locationSizeCategory: string;
  city: string;
  country: string;
}

interface AnalysisResult {
  successScore: number;
  riskLevel: "Low" | "Medium" | "High";
  marketDemand: "Low" | "Medium" | "High" | "Very High";
  competitionLevel: "Low" | "Medium" | "High" | "Very High";
  estimatedRevenueMin: number;
  estimatedRevenueMax: number;
  explanation: string;
  marketAnalysis: string;
  pricingStrategy: string;
  marketingPlan: string;
  brandingSuggestions: string;
  hiringNeeds: string;
  competitors: string; // JSON
  bestLocations: string; // JSON
}

const HIGH_DEMAND_KEYWORDS = [
  "food", "restaurant", "cafe", "delivery", "health", "fitness", "gym",
  "tech", "software", "app", "digital", "online", "e-commerce", "education",
  "beauty", "salon", "retail", "fashion", "consulting", "finance", "cleaning",
  "logistics", "healthcare", "medical", "pharmacy", "grocery", "coffee", "beverage",
  "skincare", "cosmetic", "clothing", "apparel", "travel", "tourism", "training"
];

const HIGH_COMPETITION_KEYWORDS = [
  "restaurant", "coffee", "cafe", "gym", "fitness", "salon", "beauty",
  "retail", "fashion", "food", "delivery", "e-commerce", "consulting",
  "travel", "pharmacy", "clothing", "cosmetic"
];

const LOW_COMPETITION_KEYWORDS = [
  "niche", "specialized", "unique", "innovative", "patent", "proprietary",
  "exclusive", "custom", "bespoke", "artisan", "boutique", "handmade", "organic"
];

function calculateMarketDemand(input: AnalysisInput): { demand: "Low" | "Medium" | "High" | "Very High"; score: number } {
  const text = `${input.businessType} ${input.categoryName} ${input.description} ${input.productsOrServices}`.toLowerCase();
  let demandScore = 50;

  const highMatches = HIGH_DEMAND_KEYWORDS.filter(k => text.includes(k)).length;
  demandScore += highMatches * 7;

  if (input.locationSizeCategory === "metro") demandScore += 15;
  else if (input.locationSizeCategory === "large") demandScore += 10;
  else if (input.locationSizeCategory === "medium") demandScore += 5;
  else demandScore -= 5;

  // Specific demand boosters
  const bt = input.businessType.toLowerCase();
  if (bt.includes("food delivery") || bt.includes("café") || bt.includes("coffee")) demandScore += 12;
  if (bt.includes("tech") || bt.includes("saas") || bt.includes("digital")) demandScore += 10;
  if (bt.includes("healthcare") || bt.includes("clinic") || bt.includes("pharmacy")) demandScore += 8;
  if (bt.includes("education") || bt.includes("training")) demandScore += 6;

  demandScore = Math.min(100, Math.max(0, demandScore));

  let demand: "Low" | "Medium" | "High" | "Very High";
  if (demandScore >= 75) demand = "Very High";
  else if (demandScore >= 55) demand = "High";
  else if (demandScore >= 35) demand = "Medium";
  else demand = "Low";

  return { demand, score: demandScore };
}

function calculateCompetition(input: AnalysisInput): { competition: "Low" | "Medium" | "High" | "Very High"; score: number } {
  const text = `${input.businessType} ${input.categoryName} ${input.description} ${input.productsOrServices}`.toLowerCase();
  let competitionScore = 50;

  const highMatches = HIGH_COMPETITION_KEYWORDS.filter(k => text.includes(k)).length;
  competitionScore += highMatches * 6;

  const lowMatches = LOW_COMPETITION_KEYWORDS.filter(k => text.includes(k)).length;
  competitionScore -= lowMatches * 12;

  if (input.locationSizeCategory === "metro") competitionScore += 15;
  else if (input.locationSizeCategory === "large") competitionScore += 8;
  else if (input.locationSizeCategory === "small") competitionScore -= 10;

  competitionScore = Math.min(100, Math.max(0, competitionScore));

  let competition: "Low" | "Medium" | "High" | "Very High";
  if (competitionScore >= 75) competition = "Very High";
  else if (competitionScore >= 55) competition = "High";
  else if (competitionScore >= 35) competition = "Medium";
  else competition = "Low";

  return { competition, score: competitionScore };
}

function estimateRevenue(budget: number, demandScore: number, competitionScore: number, businessType: string): { min: number; max: number } {
  const demandMultiplier = demandScore / 50;
  const competitionPenalty = (100 - competitionScore) / 100;

  let baseRate = 0.15;
  const bt = businessType.toLowerCase();
  if (bt.includes("saas") || bt.includes("tech") || bt.includes("software")) baseRate = 0.25;
  else if (bt.includes("consulting") || bt.includes("real estate")) baseRate = 0.20;
  else if (bt.includes("restaurant") || bt.includes("café") || bt.includes("food")) baseRate = 0.12;
  else if (bt.includes("logistics") || bt.includes("delivery")) baseRate = 0.10;

  const baseMonthlyRevenue = budget * baseRate * demandMultiplier * competitionPenalty;
  const min = Math.round(Math.max(500, baseMonthlyRevenue * 0.7) / 100) * 100;
  const max = Math.round(Math.max(1000, baseMonthlyRevenue * 1.4) / 100) * 100;

  return { min, max };
}

export function runAnalysis(input: AnalysisInput): AnalysisResult {
  const { demand, score: demandScore } = calculateMarketDemand(input);
  const { competition, score: competitionScore } = calculateCompetition(input);

  let successScore = 40;
  successScore += (demandScore - 50) * 0.4;
  successScore -= (competitionScore - 50) * 0.3;

  if (input.budget >= 100000) successScore += 20;
  else if (input.budget >= 50000) successScore += 12;
  else if (input.budget >= 20000) successScore += 6;
  else if (input.budget < 5000) successScore -= 10;

  if (input.locationSizeCategory === "metro") successScore += 5;
  else if (input.locationSizeCategory === "small") successScore -= 5;

  const descLength = (input.description + input.productsOrServices).length;
  if (descLength > 200) successScore += 5;
  else if (descLength < 50) successScore -= 5;

  // Products/services specificity bonus
  if (input.productsOrServices.length > 20) successScore += 4;

  successScore = Math.round(Math.min(97, Math.max(12, successScore)));

  let riskLevel: "Low" | "Medium" | "High";
  if (successScore >= 70) riskLevel = "Low";
  else if (successScore >= 45) riskLevel = "Medium";
  else riskLevel = "High";

  const { min, max } = estimateRevenue(input.budget, demandScore, competitionScore, input.businessType);

  // Get competitors and city data
  const competitorList = getCompetitorsForBusiness(input.businessType, input.categoryName);
  const cityInfo = getCityData(input.city);
  const bestZones = cityInfo ? getZonesByBusinessType(cityInfo.zones, input.businessType) : [];

  const locationPayload = cityInfo
    ? { cityLat: cityInfo.lat, cityLng: cityInfo.lng, cityName: cityInfo.city, zones: bestZones }
    : { cityLat: 25.2048, cityLng: 55.2708, cityName: input.city, zones: bestZones };

  const explanation = generateExplanation(input, successScore, riskLevel, demand, competition);
  const marketAnalysis = generateMarketAnalysis(input, demand, competition, competitorList);
  const pricingStrategy = generatePricingStrategy(input, competition);
  const marketingPlan = generateMarketingPlan(input, demand);
  const brandingSuggestions = generateBrandingSuggestions(input);
  const hiringNeeds = generateHiringNeeds(input, input.budget);

  return {
    successScore,
    riskLevel,
    marketDemand: demand,
    competitionLevel: competition,
    estimatedRevenueMin: min,
    estimatedRevenueMax: max,
    explanation,
    marketAnalysis,
    pricingStrategy,
    marketingPlan,
    brandingSuggestions,
    hiringNeeds,
    competitors: JSON.stringify(competitorList),
    bestLocations: JSON.stringify(locationPayload),
  };
}

function generateExplanation(
  input: AnalysisInput,
  score: number,
  risk: string,
  demand: string,
  competition: string
): string {
  const city = input.city || "your target city";
  const country = input.country ? `, ${input.country}` : "";
  const budgetStr = input.budget >= 1000000
    ? `$${(input.budget / 1000000).toFixed(1)}M`
    : input.budget >= 1000
    ? `$${Math.round(input.budget / 1000)}K`
    : `$${input.budget}`;

  const outlook = score >= 70 ? "strong" : score >= 50 ? "moderate" : "challenging";
  const riskNote = risk === "Low"
    ? "Your budget level and market positioning significantly minimize financial exposure."
    : risk === "Medium"
    ? "Moderate risk factors exist — careful launch planning and phased investment will be key."
    : "Elevated risk profile requires strategic mitigation: tighter cost control, phased rollout, and strong differentiation.";

  const productsNote = input.productsOrServices
    ? ` Your core offering — ${input.productsOrServices} — targets a ${demand.toLowerCase()} demand segment.`
    : "";

  return `${input.businessType} shows ${outlook} viability in the ${input.categoryName} sector within ${city}${country}. ` +
    `With a capital allocation of ${budgetStr}, market demand is ${demand.toLowerCase()} and competitive pressure is ${competition.toLowerCase()}. ` +
    `The feasibility score of ${score}/100 reflects these combined market dynamics.${productsNote} ` +
    `${riskNote} ` +
    `Success hinges on differentiated positioning, early brand trust-building, ` +
    `and precise alignment with your primary segment: ${input.targetAudience}.`;
}

function generateMarketAnalysis(
  input: AnalysisInput,
  demand: string,
  competition: string,
  competitors: Competitor[]
): string {
  const city = input.city || "the selected location";
  const locationNote = input.locationSizeCategory === "metro"
    ? `${city} is a major metropolitan market offering a massive addressable customer base, but also attracting well-funded established players.`
    : input.locationSizeCategory === "large"
    ? `${city} provides sufficient population density and consumer spending capacity for sustainable business growth.`
    : `${city} presents a focused opportunity with less saturation — ideal for building a loyal core base before scaling.`;

  const competitionNote = (competition === "Low" || competition === "Medium")
    ? "a favorable window exists to capture market share before the space matures"
    : "strong differentiation and execution speed are critical to carve out a defensible position";

  const topCompetitor = competitors[0];
  const competitorInsight = topCompetitor
    ? `The dominant player, ${topCompetitor.name} (${topCompetitor.type}), is strong in ${topCompetitor.strength.split(",")[0].toLowerCase()} — but exposed in ${topCompetitor.weakness.split(",")[0].toLowerCase()}, which is your entry opportunity.`
    : "";

  const productsContext = input.productsOrServices
    ? ` Consumer appetite for ${input.productsOrServices} in this market is driven by evolving lifestyle preferences and rising disposable income.`
    : "";

  return `The ${input.categoryName} sector in ${city} demonstrates ${demand.toLowerCase()} consumer demand. ${locationNote} ` +
    `Competition intensity is ${competition.toLowerCase()}, indicating that ${competitionNote}. ` +
    `${competitorInsight}${productsContext} ` +
    `Your target segment — "${input.targetAudience}" — aligns with documented spending growth patterns ` +
    `in this vertical, suggesting real and monetizable demand.`;
}

function generatePricingStrategy(input: AnalysisInput, competition: string): string {
  const isHighBudget = input.budget >= 50000;
  const isPremiumCategory = ["technology", "consulting", "healthcare", "luxury", "real estate"].some(c =>
    input.categoryName.toLowerCase().includes(c) || input.businessType.toLowerCase().includes(c)
  );

  const productsRef = input.productsOrServices ? `for your ${input.productsOrServices}` : "";

  if (isPremiumCategory || isHighBudget) {
    return `Adopt a value-based pricing model ${productsRef} that positions ${input.businessType} in the premium tier. ` +
      `Price 15–30% above market average to signal quality, expertise, and exclusivity. ` +
      `Structure tiered service packages (Essential, Professional, Enterprise) to capture multiple segments simultaneously. ` +
      `Introductory pricing for the first 90 days can accelerate initial acquisition without permanently anchoring price expectations downward. ` +
      `Target lifetime value (LTV) should exceed customer acquisition cost (CAC) by at least 3:1 within month 6.`;
  }

  if (competition === "High" || competition === "Very High") {
    return `Given the competitive landscape ${productsRef}, a penetration-pricing approach is recommended at launch. ` +
      `Open 10–15% below the prevailing market rate to generate initial traction and word-of-mouth. ` +
      `Bundle complementary products/services to increase perceived value while protecting unit margins. ` +
      `After 90 days, begin incremental price normalization as your brand recognition builds. ` +
      `Loyalty and subscription programs are critical here — they lock in recurring revenue while reducing churn risk in price-sensitive markets.`;
  }

  return `A market-rate pricing strategy is appropriate for ${productsRef || "your offering"}. ` +
    `Price competitively at category benchmarks while highlighting specific differentiators. ` +
    `Consider subscription or membership models to generate predictable monthly recurring revenue. ` +
    `Dynamic seasonal pricing during peak periods can improve yield without eroding everyday price perception. ` +
    `Build in a clear upsell path — at least one complementary add-on that increases average order value by 20–35%.`;
}

function generateMarketingPlan(input: AnalysisInput, _demand: string): string {
  const productsRef = input.productsOrServices ? `your ${input.productsOrServices}` : `${input.businessType}`;
  const city = input.city || "your target city";

  const channels = input.targetAudience.toLowerCase().includes("student") || input.targetAudience.toLowerCase().includes("young")
    ? "TikTok and Instagram Reels"
    : input.targetAudience.toLowerCase().includes("b2b") || input.targetAudience.toLowerCase().includes("professional")
    ? "LinkedIn and Google Search"
    : input.targetAudience.toLowerCase().includes("senior") || input.targetAudience.toLowerCase().includes("family")
    ? "Facebook and Google Maps"
    : "Instagram, Google, and TikTok";

  return `Phase 1 — Launch (Months 1–2): Build a professional presence. Claim and optimize your Google Business profile for ${city}. ` +
    `Activate ${channels} with content showcasing ${productsRef}. Launch a referral program with a clear incentive (e.g., 15% off next purchase). ` +
    `Run a geo-targeted opening campaign covering a 5km radius around your primary location. ` +
    `Phase 2 — Growth (Months 3–6): Scale paid channels with a budget of 8–12% of monthly revenue on Meta/Google Ads. ` +
    `Partner with 3–5 local micro-influencers (10k–100k followers) in the ${input.categoryName} space for authentic endorsements. ` +
    `Launch weekly educational or behind-the-scenes content to build brand trust and category authority. ` +
    `Phase 3 — Retention (Month 6+): Implement a CRM to track customer lifetime value and purchase patterns. ` +
    `Introduce a loyalty rewards program with tiered benefits. ` +
    `Collect 50+ verified reviews on Google by month 6 — reviews are the #1 local discovery signal. ` +
    `Retarget website visitors with personalized offers and abandoned-cart sequences (for e-commerce).`;
}

function generateBrandingSuggestions(input: AnalysisInput): string {
  const positioning = input.budget >= 100000 ? "premium, trustworthy market leader" : "approachable, value-forward community brand";
  const productsRef = input.productsOrServices ? `around ${input.productsOrServices}` : "";

  return `Brand Identity: Position ${input.businessType} ${productsRef} as a ${positioning} in the ${input.categoryName} space. ` +
    `Visual System: Design a clean, memorable logo using 2–3 brand colors tied to your category's emotional cues ` +
    `(trust = blue/green, energy = orange/red, luxury = black/gold). ` +
    `Typography: pair a bold display font for headlines with a readable sans-serif for body copy. ` +
    `Brand Voice: Communicate with confident warmth — knowledgeable but human. ` +
    `Your audience "${input.targetAudience}" responds to messaging that emphasizes ${
      input.targetAudience.toLowerCase().includes("luxury") ? "prestige, exclusivity, and craft" :
      input.targetAudience.toLowerCase().includes("student") ? "value, speed, and community" :
      input.targetAudience.toLowerCase().includes("professional") ? "efficiency, credibility, and ROI" :
      "reliability, quality, and belonging"
    }. ` +
    `Differentiators: Lead with your 3 strongest unique selling points — superior product quality, localized expertise, ` +
    `exceptional service — and be consistent across all touchpoints: storefront, packaging, digital channels, and staff.`;
}

function generateHiringNeeds(input: AnalysisInput, budget: number): string {
  const isLargeOperation = budget >= 100000;
  const isTech = input.businessType.toLowerCase().includes("tech") ||
    input.businessType.toLowerCase().includes("software") ||
    input.businessType.toLowerCase().includes("saas") ||
    input.categoryName.toLowerCase().includes("tech");

  const isFood = input.businessType.toLowerCase().includes("restaurant") ||
    input.businessType.toLowerCase().includes("café") ||
    input.businessType.toLowerCase().includes("food");

  const isHealthcare = input.businessType.toLowerCase().includes("clinic") ||
    input.businessType.toLowerCase().includes("health") ||
    input.categoryName.toLowerCase().includes("health");

  let coreRoles: string;
  let specialtyRoles = "";

  if (isFood) {
    coreRoles = "Head Chef / Kitchen Manager, Floor Manager, 2–4 Service Staff, Delivery Coordinator";
    specialtyRoles = " Consider a Social Media Manager specifically for food content — visual F&B content drives 40% higher engagement on Instagram.";
  } else if (isHealthcare) {
    coreRoles = "Clinic Director, 1–2 Specialist Practitioners, Receptionist & Scheduling Coordinator, Billing Specialist";
    specialtyRoles = " You'll also need a Compliance Officer familiar with local health authority regulations from day one.";
  } else if (isTech) {
    coreRoles = "Product Manager, 2 Full-Stack Developers, UX/UI Designer, DevOps Engineer";
    specialtyRoles = " A dedicated Customer Success Manager will be critical post-launch to reduce churn and gather product feedback.";
  } else if (isLargeOperation) {
    coreRoles = "Operations Manager, Sales Lead, Marketing Coordinator, Customer Service Team (2–3 staff)";
  } else {
    coreRoles = "Generalist Operations Lead, Part-time Sales/Customer Support, Social Media Coordinator";
  }

  return `Priority hires (first 3 months): ${coreRoles}.${specialtyRoles} ` +
    `Scale-up hires (months 4–12): As revenue stabilizes above break-even, add a dedicated ${input.categoryName} ` +
    `specialist, a Financial Analyst for reporting and forecasting, and an HR Coordinator once your team exceeds 8–10 people. ` +
    `Compensation strategy: Offer base salaries 5–10% above local market rate to attract quality talent, ` +
    `plus performance incentives tied to revenue milestones. For specialized roles, ` +
    `consider project-based contractors (graphic design, legal, accounting) in the first 6 months to preserve runway. ` +
    `Build a culture of accountability and clear KPIs from day one — it's significantly harder to retrofit systems later.`;
}

File: project/backend/services/analysis_engine.py
import json
from dataclasses import dataclass
from data.competitors import get_competitors_for_business
from data.city_data import get_city_data, get_zones_by_business_type

HIGH_DEMAND_KEYWORDS = [
    "food", "restaurant", "cafe", "delivery", "health", "fitness", "gym",
    "tech", "software", "app", "digital", "online", "e-commerce", "education",
    "beauty", "salon", "retail", "fashion", "consulting", "finance", "cleaning",
    "logistics", "healthcare", "medical", "pharmacy", "grocery", "coffee", "beverage",
    "skincare", "cosmetic", "clothing", "apparel", "travel", "tourism", "training",
]

HIGH_COMPETITION_KEYWORDS = [
    "restaurant", "coffee", "cafe", "gym", "fitness", "salon", "beauty",
    "retail", "fashion", "food", "delivery", "e-commerce", "consulting",
    "travel", "pharmacy", "clothing", "cosmetic",
]

LOW_COMPETITION_KEYWORDS = [
    "niche", "specialized", "unique", "innovative", "patent", "proprietary",
    "exclusive", "custom", "bespoke", "artisan", "boutique", "handmade", "organic",
]


@dataclass
class AnalysisInput:
    businessType: str
    description: str
    budget: float
    targetAudience: str
    productsOrServices: str
    categoryName: str
    locationSizeCategory: str
    city: str
    country: str


@dataclass
class AnalysisResult:
    successScore: int
    riskLevel: str
    marketDemand: str
    competitionLevel: str
    estimatedRevenueMin: int
    estimatedRevenueMax: int
    explanation: str
    marketAnalysis: str
    pricingStrategy: str
    marketingPlan: str
    brandingSuggestions: str
    hiringNeeds: str
    competitors: str
    bestLocations: str


def _calculate_market_demand(inp: AnalysisInput) -> tuple[str, int]:
    text = f"{inp.businessType} {inp.categoryName} {inp.description} {inp.productsOrServices}".lower()
    score = 50

    high_matches = sum(1 for k in HIGH_DEMAND_KEYWORDS if k in text)
    score += high_matches * 7

    if inp.locationSizeCategory == "metro":
        score += 15
    elif inp.locationSizeCategory == "large":
        score += 10
    elif inp.locationSizeCategory == "medium":
        score += 5
    else:
        score -= 5

    bt = inp.businessType.lower()
    if "food delivery" in bt or "café" in bt or "coffee" in bt:
        score += 12
    if "tech" in bt or "saas" in bt or "digital" in bt:
        score += 10
    if "healthcare" in bt or "clinic" in bt or "pharmacy" in bt:
        score += 8
    if "education" in bt or "training" in bt:
        score += 6

    score = min(100, max(0, score))

    if score >= 75:
        demand = "Very High"
    elif score >= 55:
        demand = "High"
    elif score >= 35:
        demand = "Medium"
    else:
        demand = "Low"

    return demand, score


def _calculate_competition(inp: AnalysisInput) -> tuple[str, int]:
    text = f"{inp.businessType} {inp.categoryName} {inp.description} {inp.productsOrServices}".lower()
    score = 50

    high_matches = sum(1 for k in HIGH_COMPETITION_KEYWORDS if k in text)
    score += high_matches * 6

    low_matches = sum(1 for k in LOW_COMPETITION_KEYWORDS if k in text)
    score -= low_matches * 12

    if inp.locationSizeCategory == "metro":
        score += 15
    elif inp.locationSizeCategory == "large":
        score += 8
    elif inp.locationSizeCategory == "small":
        score -= 10

    score = min(100, max(0, score))

    if score >= 75:
        competition = "Very High"
    elif score >= 55:
        competition = "High"
    elif score >= 35:
        competition = "Medium"
    else:
        competition = "Low"

    return competition, score


def _estimate_revenue(budget: float, demand_score: int, competition_score: int, business_type: str) -> tuple[int, int]:
    demand_multiplier = demand_score / 50
    competition_penalty = (100 - competition_score) / 100

    base_rate = 0.15
    bt = business_type.lower()
    if "saas" in bt or "tech" in bt or "software" in bt:
        base_rate = 0.25
    elif "consulting" in bt or "real estate" in bt:
        base_rate = 0.20
    elif "restaurant" in bt or "café" in bt or "food" in bt:
        base_rate = 0.12
    elif "logistics" in bt or "delivery" in bt:
        base_rate = 0.10

    base_monthly = budget * base_rate * demand_multiplier * competition_penalty
    min_rev = round(max(500, base_monthly * 0.7) / 100) * 100
    max_rev = round(max(1000, base_monthly * 1.4) / 100) * 100

    return int(min_rev), int(max_rev)


def _generate_explanation(inp: AnalysisInput, score: int, risk: str, demand: str, competition: str) -> str:
    city = inp.city or "your target city"
    country = f", {inp.country}" if inp.country else ""

    if inp.budget >= 1_000_000:
        budget_str = f"${inp.budget / 1_000_000:.1f}M"
    elif inp.budget >= 1000:
        budget_str = f"${round(inp.budget / 1000)}K"
    else:
        budget_str = f"${inp.budget:.0f}"

    outlook = "strong" if score >= 70 else "moderate" if score >= 50 else "challenging"

    if risk == "Low":
        risk_note = "Your budget level and market positioning significantly minimize financial exposure."
    elif risk == "Medium":
        risk_note = "Moderate risk factors exist — careful launch planning and phased investment will be key."
    else:
        risk_note = "Elevated risk profile requires strategic mitigation: tighter cost control, phased rollout, and strong differentiation."

    products_note = f" Your core offering — {inp.productsOrServices} — targets a {demand.lower()} demand segment." if inp.productsOrServices else ""

    return (
        f"{inp.businessType} shows {outlook} viability in the {inp.categoryName} sector within {city}{country}. "
        f"With a capital allocation of {budget_str}, market demand is {demand.lower()} and competitive pressure is {competition.lower()}. "
        f"The feasibility score of {score}/100 reflects these combined market dynamics.{products_note} "
        f"{risk_note} "
        f"Success hinges on differentiated positioning, early brand trust-building, "
        f"and precise alignment with your primary segment: {inp.targetAudience}."
    )


def _generate_market_analysis(inp: AnalysisInput, demand: str, competition: str, competitors: list) -> str:
    city = inp.city or "the selected location"

    if inp.locationSizeCategory == "metro":
        location_note = f"{city} is a major metropolitan market offering a massive addressable customer base, but also attracting well-funded established players."
    elif inp.locationSizeCategory == "large":
        location_note = f"{city} provides sufficient population density and consumer spending capacity for sustainable business growth."
    else:
        location_note = f"{city} presents a focused opportunity with less saturation — ideal for building a loyal core base before scaling."

    if competition in ("Low", "Medium"):
        competition_note = "a favorable window exists to capture market share before the space matures"
    else:
        competition_note = "strong differentiation and execution speed are critical to carve out a defensible position"

    competitor_insight = ""
    if competitors:
        top = competitors[0]
        competitor_insight = f"The dominant player, {top['name']} ({top['type']}), is strong in {top['strength'].split(',')[0].lower()} — but exposed in {top['weakness'].split(',')[0].lower()}, which is your entry opportunity."

    products_context = f" Consumer appetite for {inp.productsOrServices} in this market is driven by evolving lifestyle preferences and rising disposable income." if inp.productsOrServices else ""

    return (
        f"The {inp.categoryName} sector in {city} demonstrates {demand.lower()} consumer demand. {location_note} "
        f"Competition intensity is {competition.lower()}, indicating that {competition_note}. "
        f"{competitor_insight}{products_context} "
        f"Your target segment — \"{inp.targetAudience}\" — aligns with documented spending growth patterns "
        f"in this vertical, suggesting real and monetizable demand."
    )


def _generate_pricing_strategy(inp: AnalysisInput, competition: str) -> str:
    is_high_budget = inp.budget >= 50000
    premium_keywords = ["technology", "consulting", "healthcare", "luxury", "real estate"]
    is_premium = any(k in inp.categoryName.lower() or k in inp.businessType.lower() for k in premium_keywords)
    products_ref = f"for your {inp.productsOrServices}" if inp.productsOrServices else ""

    if is_premium or is_high_budget:
        return (
            f"Adopt a value-based pricing model {products_ref} that positions {inp.businessType} in the premium tier. "
            f"Price 15–30% above market average to signal quality, expertise, and exclusivity. "
            f"Structure tiered service packages (Essential, Professional, Enterprise) to capture multiple segments simultaneously. "
            f"Introductory pricing for the first 90 days can accelerate initial acquisition without permanently anchoring price expectations downward. "
            f"Target lifetime value (LTV) should exceed customer acquisition cost (CAC) by at least 3:1 within month 6."
        )

    if competition in ("High", "Very High"):
        return (
            f"Given the competitive landscape {products_ref}, a penetration-pricing approach is recommended at launch. "
            f"Open 10–15% below the prevailing market rate to generate initial traction and word-of-mouth. "
            f"Bundle complementary products/services to increase perceived value while protecting unit margins. "
            f"After 90 days, begin incremental price normalization as your brand recognition builds. "
            f"Loyalty and subscription programs are critical here — they lock in recurring revenue while reducing churn risk in price-sensitive markets."
        )

    return (
        f"A market-rate pricing strategy is appropriate for {products_ref or 'your offering'}. "
        f"Price competitively at category benchmarks while highlighting specific differentiators. "
        f"Consider subscription or membership models to generate predictable monthly recurring revenue. "
        f"Dynamic seasonal pricing during peak periods can improve yield without eroding everyday price perception. "
        f"Build in a clear upsell path — at least one complementary add-on that increases average order value by 20–35%."
    )


def _generate_marketing_plan(inp: AnalysisInput) -> str:
    products_ref = f"your {inp.productsOrServices}" if inp.productsOrServices else inp.businessType
    city = inp.city or "your target city"
    ta_lower = inp.targetAudience.lower()

    if "student" in ta_lower or "young" in ta_lower:
        channels = "TikTok and Instagram Reels"
    elif "b2b" in ta_lower or "professional" in ta_lower:
        channels = "LinkedIn and Google Search"
    elif "senior" in ta_lower or "family" in ta_lower:
        channels = "Facebook and Google Maps"
    else:
        channels = "Instagram, Google, and TikTok"

    return (
        f"Phase 1 — Launch (Months 1–2): Build a professional presence. Claim and optimize your Google Business profile for {city}. "
        f"Activate {channels} with content showcasing {products_ref}. Launch a referral program with a clear incentive (e.g., 15% off next purchase). "
        f"Run a geo-targeted opening campaign covering a 5km radius around your primary location. "
        f"Phase 2 — Growth (Months 3–6): Scale paid channels with a budget of 8–12% of monthly revenue on Meta/Google Ads. "
        f"Partner with 3–5 local micro-influencers (10k–100k followers) in the {inp.categoryName} space for authentic endorsements. "
        f"Launch weekly educational or behind-the-scenes content to build brand trust and category authority. "
        f"Phase 3 — Retention (Month 6+): Implement a CRM to track customer lifetime value and purchase patterns. "
        f"Introduce a loyalty rewards program with tiered benefits. "
        f"Collect 50+ verified reviews on Google by month 6 — reviews are the #1 local discovery signal. "
        f"Retarget website visitors with personalized offers and abandoned-cart sequences (for e-commerce)."
    )


def _generate_branding(inp: AnalysisInput) -> str:
    positioning = "premium, trustworthy market leader" if inp.budget >= 100000 else "approachable, value-forward community brand"
    products_ref = f"around {inp.productsOrServices}" if inp.productsOrServices else ""
    ta_lower = inp.targetAudience.lower()

    if "luxury" in ta_lower:
        msg_focus = "prestige, exclusivity, and craft"
    elif "student" in ta_lower:
        msg_focus = "value, speed, and community"
    elif "professional" in ta_lower:
        msg_focus = "efficiency, credibility, and ROI"
    else:
        msg_focus = "reliability, quality, and belonging"

    return (
        f"Brand Identity: Position {inp.businessType} {products_ref} as a {positioning} in the {inp.categoryName} space. "
        f"Visual System: Design a clean, memorable logo using 2–3 brand colors tied to your category's emotional cues "
        f"(trust = blue/green, energy = orange/red, luxury = black/gold). "
        f"Typography: pair a bold display font for headlines with a readable sans-serif for body copy. "
        f"Brand Voice: Communicate with confident warmth — knowledgeable but human. "
        f"Your audience \"{inp.targetAudience}\" responds to messaging that emphasizes {msg_focus}. "
        f"Differentiators: Lead with your 3 strongest unique selling points — superior product quality, localized expertise, "
        f"exceptional service — and be consistent across all touchpoints: storefront, packaging, digital channels, and staff."
    )


def _generate_hiring(inp: AnalysisInput) -> str:
    bt = inp.businessType.lower()
    cat = inp.categoryName.lower()
    is_large = inp.budget >= 100000
    is_tech = any(k in bt or k in cat for k in ["tech", "software", "saas"])
    is_food = any(k in bt for k in ["restaurant", "café", "food"])
    is_health = any(k in bt or k in cat for k in ["clinic", "health"])

    if is_food:
        core_roles = "Head Chef / Kitchen Manager, Floor Manager, 2–4 Service Staff, Delivery Coordinator"
        specialty = " Consider a Social Media Manager specifically for food content — visual F&B content drives 40% higher engagement on Instagram."
    elif is_health:
        core_roles = "Clinic Director, 1–2 Specialist Practitioners, Receptionist & Scheduling Coordinator, Billing Specialist"
        specialty = " You'll also need a Compliance Officer familiar with local health authority regulations from day one."
    elif is_tech:
        core_roles = "Product Manager, 2 Full-Stack Developers, UX/UI Designer, DevOps Engineer"
        specialty = " A dedicated Customer Success Manager will be critical post-launch to reduce churn and gather product feedback."
    elif is_large:
        core_roles = "Operations Manager, Sales Lead, Marketing Coordinator, Customer Service Team (2–3 staff)"
        specialty = ""
    else:
        core_roles = "Generalist Operations Lead, Part-time Sales/Customer Support, Social Media Coordinator"
        specialty = ""

    return (
        f"Priority hires (first 3 months): {core_roles}.{specialty} "
        f"Scale-up hires (months 4–12): As revenue stabilizes above break-even, add a dedicated {inp.categoryName} "
        f"specialist, a Financial Analyst for reporting and forecasting, and an HR Coordinator once your team exceeds 8–10 people. "
        f"Compensation strategy: Offer base salaries 5–10% above local market rate to attract quality talent, "
        f"plus performance incentives tied to revenue milestones. For specialized roles, "
        f"consider project-based contractors (graphic design, legal, accounting) in the first 6 months to preserve runway. "
        f"Build a culture of accountability and clear KPIs from day one — it's significantly harder to retrofit systems later."
    )


def run_analysis(inp: AnalysisInput) -> AnalysisResult:
    demand, demand_score = _calculate_market_demand(inp)
    competition, competition_score = _calculate_competition(inp)

    success_score = 40
    success_score += (demand_score - 50) * 0.4
    success_score -= (competition_score - 50) * 0.3

    if inp.budget >= 100000:
        success_score += 20
    elif inp.budget >= 50000:
        success_score += 12
    elif inp.budget >= 20000:
        success_score += 6
    elif inp.budget < 5000:
        success_score -= 10

    if inp.locationSizeCategory == "metro":
        success_score += 5
    elif inp.locationSizeCategory == "small":
        success_score -= 5

    desc_len = len(inp.description + inp.productsOrServices)
    if desc_len > 200:
        success_score += 5
    elif desc_len < 50:
        success_score -= 5

    if len(inp.productsOrServices) > 20:
        success_score += 4

    success_score = int(min(97, max(12, round(success_score))))

    if success_score >= 70:
        risk_level = "Low"
    elif success_score >= 45:
        risk_level = "Medium"
    else:
        risk_level = "High"

    rev_min, rev_max = _estimate_revenue(inp.budget, demand_score, competition_score, inp.businessType)

    competitors = get_competitors_for_business(inp.businessType, inp.categoryName)
    city_info = get_city_data(inp.city)

    if city_info:
        best_zones = get_zones_by_business_type(city_info["zones"], inp.businessType)
        location_payload = {
            "cityLat": city_info["lat"],
            "cityLng": city_info["lng"],
            "cityName": city_info["city"],
            "zones": best_zones,
        }
    else:
        location_payload = {
            "cityLat": 25.2048,
            "cityLng": 55.2708,
            "cityName": inp.city,
            "zones": [],
        }

    return AnalysisResult(
        successScore=success_score,
        riskLevel=risk_level,
        marketDemand=demand,
        competitionLevel=competition,
        estimatedRevenueMin=rev_min,
        estimatedRevenueMax=rev_max,
        explanation=_generate_explanation(inp, success_score, risk_level, demand, competition),
        marketAnalysis=_generate_market_analysis(inp, demand, competition, competitors),
        pricingStrategy=_generate_pricing_strategy(inp, competition),
        marketingPlan=_generate_marketing_plan(inp),
        brandingSuggestions=_generate_branding(inp),
        hiringNeeds=_generate_hiring(inp),
        competitors=json.dumps(competitors),
        bestLocations=json.dumps(location_payload),
    )
